homeassistant/components/shelly/config_flow.py

```diff
@@ -0,0 +1,157 @@
+    """Return Shelly devices connected to the MQTT broker."""
+    _LOGGER.debug("Starting Shelly discovery...")
+    mqtt = hass.components.mqtt
+
+    devices = []
+
+    def message_received(msg):
+        devices.append(json.loads(msg.payload))
+
+    unsubscribe = await mqtt.async_subscribe(COMMON_ANNOUNCE_TOPIC, message_received)
+    mqtt.async_publish(COMMON_COMMAND_TOPIC, ANNOUNCE_COMMAND)
+    # give all devices 5 seconds to announce themselves
+    await asyncio.sleep(timeout)
+    unsubscribe()
+
+    return devices
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Shelly."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
+
+    def __init__(self):
+        """Initialize the Shelly config flow."""
+        self._device = None
+        self._devices = []
+
+    async def async_step_user(self, user_input=None):
+        """Handle user flow."""
+        if "mqtt" not in self.hass.config.components:
+            return self.async_abort(reason="mqtt_required")
+
+        if user_input is not None:
+            return await self.async_step_device()
+
+        return self.async_show_form(step_id="user")
+
+    async def async_step_device(self, user_input=None):
+        """Let the user pick the device."""
+        if user_input:
+            for device in self._devices:
+                if device[CONF_ID] == user_input[CONF_DEVICE_ID]:
+                    self._device = device[CONF_ID]
+                    await self.async_set_unique_id(self._device)
+                    self._abort_if_unique_id_configured()
+                    return await self.async_step_topic()
+
+        discovery = await async_discovery(self.hass)
+        for device in discovery:
+            configured = any(
+                entry.unique_id == device[CONF_ID]
+                for entry in self._async_current_entries()
+            )
+
+            if not configured:
+                self._devices.append(device)
+
+        _LOGGER.debug("Discovered Shelly devices %s", pformat(self._devices))
+
+        if self._devices:
+            names = [device[CONF_ID] for device in self._devices]
+
+            return self.async_show_form(
+                step_id="device",
+                data_schema=vol.Schema({vol.Required(CONF_DEVICE_ID): vol.In(names)}),
+            )
+
+        return self.async_abort(reason="no_devices_found")
+
+    async def async_step_topic(self, user_input=None):
+        """Model and topic configuration for Shelly device."""
+        errors = {}
+        if user_input:
+            try:
+                self.hass.components.mqtt.valid_publish_topic(user_input[CONF_TOPIC])
+                success = await validate_input(self.hass, user_input)
+                if success:
+                    device_id = self._device
+                    model = device_id.split("-")[0]
+                    title = f"{MODELS[model][MODEL_TITLE]} ({device_id})"
```

Comment:
This line will cause `KeyError` if device is not supported, e.g. Shelly Plug S.

Reply:
Good catch, thanks!